Security issue: upgrade prost depedency to 0.8.0

I run `cargo audit` against a project of mine, Adding the `opentelemetry-otlp` dependency lead to this discovery:

```
Crate:         prost-types
Version:       0.7.0
Title:         Conversion from `prost_types::Timestamp` to `SystemTime` can cause an overflow and panic
Date:          2021-07-08
ID:            RUSTSEC-2021-0073
URL:           https://rustsec.org/advisories/RUSTSEC-2021-0073
Solution:      Upgrade to >=0.8.0
Dependency tree: 
prost-types 0.7.0
└── prost-build 0.7.0
    └── tonic-build 0.4.2
        └── opentelemetry-otlp 0.7.0
```

I've tried to contribute a fix for that by forking this repo and updating the prost dependency. Unfortunately prost 0.8 requires tonic 0.5, which introduces some breaking changes inside of the automatically generated code.

The breaking change happens inside of the automatically generated `TraceServiceClient`.

## Code generated with prost 0.7 and tonic 0.4

This is the relevant snippet from the `opentelemetry.proto.collector.trace.v1.rs` file that is automatically generated:

```rust
pub mod trace_service_client {
     #![allow(unused_variables, dead_code, missing_docs)]
     use tonic::codegen::*;
     #[doc = " Service that can be used to push spans between one Application instrumented with"]
     #[doc = " OpenTelemetry and an collector, or between an collector and a central collector (in this"]
     #[doc = " case spans are sent/received to/from multiple Applications)."]
     pub struct TraceServiceClient<T> {
         inner: tonic::client::Grpc<T>,
     }
     impl TraceServiceClient<tonic::transport::Channel> {
         #[doc = r" Attempt to create a new client by connecting to a given endpoint."]
         pub async fn connect<D>(dst: D) -> Result<Self, tonic::transport::Error>
         where
             D: std::convert::TryInto<tonic::transport::Endpoint>,
             D::Error: Into<StdError>,
         {
             let conn = tonic::transport::Endpoint::new(dst)?.connect().await?;
             Ok(Self::new(conn))
         }
     }
     impl<T> TraceServiceClient<T>
     where
         T: tonic::client::GrpcService<tonic::body::BoxBody>,
         T::ResponseBody: Body + HttpBody + Send + 'static,
         T::Error: Into<StdError>,
         <T::ResponseBody as HttpBody>::Error: Into<StdError> + Send,
     {
         pub fn new(inner: T) -> Self {
             let inner = tonic::client::Grpc::new(inner);
             Self { inner }
         }
         pub fn with_interceptor(inner: T, interceptor: impl Into<tonic::Interceptor>) -> Self {
             let inner = tonic::client::Grpc::with_interceptor(inner, interceptor);
             Self { inner }
         }
```

As you can see, `new` and `with_interceptor` always return an instance of `TraceServiceClient`.

## Code generated with prost 0.8 and tonic 0.5

This is the same code, generated with the latest versions of the crates:

```rust
 pub mod trace_service_client {
     #![allow(unused_variables, dead_code, missing_docs)]
     use tonic::codegen::*;
     #[doc = " Service that can be used to push spans between one Application instrumented with"]
     #[doc = " OpenTelemetry and an collector, or between an collector and a central collector (in this"]
     #[doc = " case spans are sent/received to/from multiple Applications)."]
     #[derive(Debug, Clone)]
     pub struct TraceServiceClient<T> {
         inner: tonic::client::Grpc<T>,
     }
     impl TraceServiceClient<tonic::transport::Channel> {
         #[doc = r" Attempt to create a new client by connecting to a given endpoint."]
         pub async fn connect<D>(dst: D) -> Result<Self, tonic::transport::Error>
         where
             D: std::convert::TryInto<tonic::transport::Endpoint>,
             D::Error: Into<StdError>,
         {
             let conn = tonic::transport::Endpoint::new(dst)?.connect().await?;
             Ok(Self::new(conn))
         }
     }
     impl<T> TraceServiceClient<T>
     where
         T: tonic::client::GrpcService<tonic::body::BoxBody>,
         T::ResponseBody: Body + Send + Sync + 'static,
         T::Error: Into<StdError>,
         <T::ResponseBody as Body>::Error: Into<StdError> + Send,
     {
         pub fn new(inner: T) -> Self {
             let inner = tonic::client::Grpc::new(inner);
             Self { inner }
         }
         pub fn with_interceptor<F>(
             inner: T,
             interceptor: F,
         ) -> TraceServiceClient<InterceptedService<T, F>>
         where
             F: FnMut(tonic::Request<()>) -> Result<tonic::Request<()>, tonic::Status>,
             T: Service<
                 http::Request<tonic::body::BoxBody>,
                 Response = http::Response<
                     <T as tonic::client::GrpcService<tonic::body::BoxBody>>::ResponseBody,
                 >,
             >,
             <T as Service<http::Request<tonic::body::BoxBody>>>::Error:
                 Into<StdError> + Send + Sync,
         {
             TraceServiceClient::new(InterceptedService::new(inner, interceptor))
         }
```

Now the return type of the `new` and `with_interceptor` methods is different.

That leads to the following compilation error:

```console
error[E0308]: `match` arms have incompatible types
   --> opentelemetry-otlp/src/span.rs:358:17
    |
355 |           let client = match tonic_config.metadata.to_owned() {
    |                        -------------------------------------- `match` arms have incompatible types
356 |               None => TonicTraceServiceClient::new(channel),
    |                       ------------------------------------- this is found to be of type `TraceServiceClient<Channel>`
357 |               Some(metadata) => {
358 | /                 TonicTraceServiceClient::with_interceptor(channel, move |mut req: Request<()>| {
359 | |                     for key_and_value in metadata.iter() {
360 | |                         match key_and_value {
361 | |                             KeyAndValueRef::Ascii(key, value) => {
...   |
370 | |                     Ok(req)
371 | |                 })
    | |__________________^ expected struct `Channel`, found struct `tonic::codegen::InterceptedService`
    |
    = note: expected type `TraceServiceClient<Channel>`
             found struct `TraceServiceClient<tonic::codegen::InterceptedService<Channel, [closure@opentelemetry-otlp/src/span.rs:358:68: 371:18]>>`
```

I don't know how to best solve this problem... I spent some time trying, but I couldn't find a good solution. I've to admit, I'm also not experienced with protobuf code generation

--- a/examples/external-otlp-tonic-tokio/Cargo.toml
+++ b/examples/external-otlp-tonic-tokio/Cargo.toml
@@ -9,5 +9,5 @@ opentelemetry = { path = "../../opentelemetry", features = ["rt-tokio", "metrics
 opentelemetry-otlp = { path = "../../opentelemetry-otlp", features = ["tonic", "tls", "tls-roots"] }
 serde_json = "1.0"
 tokio = { version = "1.0", features = ["full"] }
-tonic = "0.4.0"
+tonic = "0.5.0"
 url = "2.2.0"

--- a/opentelemetry-otlp/Cargo.toml
+++ b/opentelemetry-otlp/Cargo.toml
@@ -39,19 +39,22 @@ async-trait = "0.1"
 futures = "0.3"
 grpcio = { version = "0.9", optional = true }
 opentelemetry = { version = "0.15", default-features = false, features = ["trace"], path = "../opentelemetry" }
-prost = { version = "0.7", optional = true }
+prost = { version = "0.8", optional = true }
 protobuf = { version = "2.18", optional = true }
 thiserror = "1.0"
-tonic = { version = "0.4", optional = true }
+tonic = { version = "0.5", optional = true }
 tokio = { version = "1.0", features = ["full"], optional = true }
 opentelemetry-http = { version = "0.4", path = "../opentelemetry-http", optional = true }
 reqwest = { version = "0.11", optional = true, default-features = false }
 surf = { version = "2.0", optional = true, default-features = false }
 http = "0.2"
+tower = { version = "0.4", optional = true }
 
 [dev-dependencies]
 chrono = "0.4"
 tokio-stream = { version = "0.1", features = ["net"] }
+# need tokio runtime to run smoke tests.
+opentelemetry = { version = "0.15", features = ["trace", "rt-tokio"], path = "../opentelemetry" }
 protobuf-codegen = { version = "2.16"}
 protoc-grpcio = { version = "3.0"}
 
@@ -59,7 +62,7 @@ protoc-grpcio = { version = "3.0"}
 trace = ["opentelemetry/trace"]
 metrics = ["opentelemetry/metrics", "default"]
 
-default = ["tonic", "tonic-build", "prost", "tokio"]
+default = ["tonic", "tonic-build", "prost", "tokio", "tower"]
 tls = ["tonic/tls"]
 tls-roots = ["tls", "tonic/tls-roots"]
 
@@ -75,5 +78,5 @@ reqwest-rustls = ["reqwest", "reqwest/rustls-tls-native-roots"]
 surf-client = ["surf", "opentelemetry-http/surf"]
 
 [build-dependencies]
-tonic-build = { version = "0.4", optional = true }
-prost-build = {version = "0.7", optional = true }
+tonic-build = { version = "0.5", optional = true }
+prost-build = {version = "0.8", optional = true }

--- a/opentelemetry-otlp/Cargo.toml
+++ b/opentelemetry-otlp/Cargo.toml
@@ -48,7 +48,6 @@ opentelemetry-http = { version = "0.4", path = "../opentelemetry-http", optional
 reqwest = { version = "0.11", optional = true, default-features = false }
 surf = { version = "2.0", optional = true, default-features = false }
 http = "0.2"
-tower = { version = "0.4", optional = true }
 
 [dev-dependencies]
 chrono = "0.4"
@@ -62,7 +61,7 @@ protoc-grpcio = { version = "3.0"}
 trace = ["opentelemetry/trace"]
 metrics = ["opentelemetry/metrics", "default"]
 
-default = ["tonic", "tonic-build", "prost", "tokio", "tower"]
+default = ["tonic", "tonic-build", "prost", "tokio"]
 tls = ["tonic/tls"]
 tls-roots = ["tls", "tonic/tls-roots"]
 

--- a/opentelemetry-otlp/src/metric.rs
+++ b/opentelemetry-otlp/src/metric.rs
@@ -252,6 +252,7 @@ pub struct MetricsExporter {
     #[cfg(feature = "tokio")]
     sender: Arc<Mutex<tokio::sync::mpsc::Sender<ExportMsg>>>,
     export_kind_selector: Arc<dyn ExportKindFor + Send + Sync>,
+    metadata: Option<tonic::metadata::MetadataMap>,
 }
 
 impl Debug for MetricsExporter {
@@ -274,7 +275,7 @@ impl MetricsExporter {
     #[cfg(feature = "tonic")]
     pub fn new<T: ExportKindFor + Send + Sync + 'static>(
         config: ExportConfig,
-        tonic_config: TonicConfig,
+        mut tonic_config: TonicConfig,
         export_selector: T,
     ) -> Result<MetricsExporter> {
         let endpoint =
@@ -297,25 +298,7 @@ impl MetricsExporter {
             .connect_lazy()
             .map_err::<crate::Error, _>(Into::into)?;
 
-        let client = match tonic_config.metadata.to_owned() {
-            None => MetricsServiceClient::new(channel),
-            Some(metadata) => {
-                MetricsServiceClient::with_interceptor(channel, move |mut req: Request<()>| {
-                    for key_and_value in metadata.iter() {
-                        match key_and_value {
-                            KeyAndValueRef::Ascii(key, value) => {
-                                req.metadata_mut().append(key, value.to_owned())
-                            }
-                            KeyAndValueRef::Binary(key, value) => {
-                                req.metadata_mut().append_bin(key, value.to_owned())
-                            }
-                        };
-                    }
-
-                    Ok(req)
-                })
-            }
-        };
+        let client = MetricsServiceClient::new(channel);
 
         let (sender, mut receiver) = tokio::sync::mpsc::channel::<ExportMsg>(2);
         tokio::spawn(Box::pin(async move {
@@ -334,6 +317,7 @@ impl MetricsExporter {
         Ok(MetricsExporter {
             sender: Arc::new(Mutex::new(sender)),
             export_kind_selector: Arc::new(export_selector),
+            metadata: tonic_config.metadata.take(),
         })
     }
 }
@@ -359,7 +343,19 @@ impl Exporter for MetricsExporter {
                 Err(err) => Err(err),
             }
         })?;
-        let request = Request::new(sink(resource_metrics));
+        let mut request = Request::new(sink(resource_metrics));
+        if let Some(metadata) = &self.metadata {
+            for key_and_value in metadata.iter() {
+                match key_and_value {
+                    KeyAndValueRef::Ascii(key, value) => {
+                        request.metadata_mut().append(key, value.to_owned())
+                    }
+                    KeyAndValueRef::Binary(key, value) => {
+                        request.metadata_mut().append_bin(key, value.to_owned())
+                    }
+                };
+            }
+        }
         self.sender
             .lock()
             .map(|sender| {

--- a/opentelemetry-otlp/src/span.rs
+++ b/opentelemetry-otlp/src/span.rs
@@ -352,30 +352,10 @@ impl SpanExporter {
         tonic_config: TonicConfig,
         channel: tonic::transport::Channel,
     ) -> Result<Self, crate::Error> {
-        let client = match tonic_config.metadata.to_owned() {
-            None => TonicTraceServiceClient::new(channel),
-            Some(metadata) => {
-                TonicTraceServiceClient::with_interceptor(channel, move |mut req: Request<()>| {
-                    for key_and_value in metadata.iter() {
-                        match key_and_value {
-                            KeyAndValueRef::Ascii(key, value) => {
-                                req.metadata_mut().append(key, value.to_owned())
-                            }
-                            KeyAndValueRef::Binary(key, value) => {
-                                req.metadata_mut().append_bin(key, value.to_owned())
-                            }
-                        };
-                    }
-
-                    Ok(req)
-                })
-            }
-        };
-
         Ok(SpanExporter::Tonic {
             timeout: config.timeout,
             metadata: tonic_config.metadata,
-            trace_exporter: client,
+            trace_exporter: TonicTraceServiceClient::new(channel),
         })
     }
 
@@ -466,10 +446,27 @@ impl opentelemetry::sdk::export::trace::SpanExporter for SpanExporter {
             }
 
             #[cfg(feature = "tonic")]
-            SpanExporter::Tonic { trace_exporter, .. } => {
-                let request = Request::new(TonicRequest {
+            SpanExporter::Tonic {
+                trace_exporter,
+                metadata,
+                ..
+            } => {
+                let mut request = Request::new(TonicRequest {
                     resource_spans: batch.into_iter().map(Into::into).collect(),
                 });
+
+                if let Some(metadata) = metadata {
+                    for key_and_value in metadata.iter() {
+                        match key_and_value {
+                            KeyAndValueRef::Ascii(key, value) => {
+                                request.metadata_mut().append(key, value.to_owned())
+                            }
+                            KeyAndValueRef::Binary(key, value) => {
+                                request.metadata_mut().append_bin(key, value.to_owned())
+                            }
+                        };
+                    }
+                }
 
                 trace_exporter
                     .to_owned()

--- a/opentelemetry-otlp/tests/smoke.rs
+++ b/opentelemetry-otlp/tests/smoke.rs
@@ -27,6 +27,11 @@ impl TraceService for MockServer {
         request: tonic::Request<ExportTraceServiceRequest>,
     ) -> Result<tonic::Response<ExportTraceServiceResponse>, tonic::Status> {
         println!("Sending request into channel...");
+        // assert we have required metadata key
+        assert_eq!(
+            request.metadata().get("x-header-key"),
+            Some(&("header-value".parse().unwrap()))
+        );
         self.tx
             .lock()
             .unwrap()
@@ -68,12 +73,15 @@ async fn smoke_tracer() {
 
     {
         println!("Installing tracer...");
+        let mut metadata = tonic::metadata::MetadataMap::new();
+        metadata.insert("x-header-key", "header-value".parse().unwrap());
         let tracer = opentelemetry_otlp::new_pipeline()
             .tracing()
             .with_exporter(
                 opentelemetry_otlp::new_exporter()
                     .tonic()
-                    .with_endpoint(format!("http://{}", addr)),
+                    .with_endpoint(format!("http://{}", addr))
+                    .with_metadata(metadata),
             )
             .install_batch(opentelemetry::runtime::Tokio)
             .expect("failed to install");

